Add OnSendToIRCMessage and OnSendToClientMessage

elyscape opened the pull request:
As described in [this comment](https://github.com/znc/znc/pull/1396#issuecomment-293179090).

elyscape commented:
@DarthGandalf Should I implement CMessage variants of `PutIRC` and `PutUser`?

DarthGandalf commented:
I see several issues:
* Any changes to the message made by modules are ignored
* Rather than parsing the line for the sake of calling modules, PutClient should accept CMessage, and stringify it to get the line to send. Also see #1193

elyscape commented:
@DarthGandalf Any recommendations for handling changes to the message made by modules? Should we make a copy of the `CMessage` before sending it to modules?

As for making a version of `PutClient` that accepts a `CMessage`, [one already exists][putclient], but ultimately [it stringifies the `CMessage`][tostring] and [calls the `CString` variant][call], which is what [calls modules and writes to the client][write]. Changing things so that this is handled by `PutClient(CMessage)` would require some more significant changes. What do you think?

[putclient]: https://github.com/znc/znc/blob/3189ce7/src/Client.cpp#L495
[tostring]: https://github.com/znc/znc/blob/3189ce7/src/Client.cpp#L566
[call]: https://github.com/znc/znc/blob/3189ce7/src/Client.cpp#L592
[write]: https://github.com/znc/znc/blob/3189ce7/src/Client.cpp#L488-L492

elyscape commented:
Oh wait, I misunderstood your first issue. Give me a moment and I'll fix that.

elyscape commented:
Actually, on review, it wasn't that difficult to change `PutClient` such that the `PutClient(CMessage)` handled actually sending messages and calling hooks.

elyscape commented:
Another bonus from doing that is that it eliminated the need for a separate module hook for specifying message tags, as that can now be done inside `OnSendToClientMessage()`.

elyscape commented:
@DarthGandalf Please take another look when you have the opportunity. Thanks!

elyscape commented on modules/modpython/znc.py:
```diff
@@ -655,6 +655,22 @@ def OnTopicMessage(self, msg):
+    def OnSendToIRCMessage(self, msg):
+        modified = String()
+        old = modified.s = msg.ToString()
+        ret = self.OnChanBufferPlayLine(modified)
```
Good catch. Will fix.

elyscape commented:
The copying behavior was based on the way [`OnChan/PrivBufferPlayMessage` is handled][onbufferplaymessage], along with the way [most of the other `OnXMessage` hooks are handled][message-hooks]. While there is a performance hit for modules that don't override `OnSendToIRCMessage` or `OnSendToClientMessage`, if those are overridden then the non-Message variants aren't even called. Then again, `OnSendToIRCMessage` and `OnSendToClientMessage` will be called much more often, so I'll switch to a method of calling both directly.

[onbufferplaymessage]: https://github.com/znc/znc/blob/f37aa30/src/Modules.cpp#L753-L772
[message-hooks]: https://github.com/znc/znc/pull/1000/files#diff-12d169eaaada02c163f80979e8ead3a1

DarthGandalf commented:
For those OnXMessage hooks, I don't know how to avoid such per-module conversion back and forth. If you have any better idea, please say.

Hopefully, after several versions, after the sufficient transition period, old deprecated hooks can be removed, and such compatibility code can be dropped.

elyscape commented:
Okay, I've switched things such that both `OnSendToClient/IRC` and `OnSendToClient/IRCMessage` are called for performance reasons. I've also added tests for the new hooks.

DarthGandalf commented on src/Client.cpp:
```diff
@@ -483,13 +483,8 @@ CString CClient::GetFullName() const {
-    DEBUG("(" << GetFullName() << ") ZNC -> CLI [" << sCopy << "]");
-    Write(sCopy + "\r\n");
+    CMessage Message(sLine);
+    PutClient(Message);
```
I like how you handled this.

DarthGandalf commented on src/IRCSock.cpp:
```diff
@@ -1154,13 +1154,20 @@ void CIRCSock::TrySend() {
         CString& sLine = m_vsSendQueue.front();
-        IRCSOCKMODULECALL(OnSendToIRC(sLine), &bSkip);
+
+        CMessage Message(sLine);
```
Probably send queue should store messages instead of lines... That's tied to #1193

elyscape replied:
Yeah, I figured that I would handle in a separate pull request.

elyscape commented on test/IRCTest.h:
```diff
@@ -188,6 +198,7 @@ class TestModule : public CModule {
     std::vector<CClient*> vClients;
     std::vector<CChan*> vChannels;
     EModRet eAction = CONTINUE;
+    bool bSendHooks = false;
```
I did this so that, when testing other hooks, the `OnSendToClient/IRCMessage` hooks don't fire. Before I put it in here, it was necessary to modify other tests to indicate that `OnSendToClient/IRCMessage` would be included in the list of hooks that fired. Additionally, it simplified handling of `HALT` for the tests.

DarthGandalf replied:
I don't see a problem with indicating full list of callbacks called.
How is HALT simplified?

elyscape replied:
As an example of the sort of thing that would happen, [the test for `OnJoinMessage`][onjoinmessage] would have had to be changed from this:
```cpp
TEST_F(IRCSockTest, OnJoinMessage) {
    CMessage msg(":somebody JOIN #chan");
    m_pTestSock->ReadLine(msg.ToString());

    EXPECT_THAT(m_pTestModule->vsHooks, ElementsAre("OnJoinMessage"));
    EXPECT_THAT(m_pTestModule->vsMessages, ElementsAre(msg.ToString()));
    EXPECT_THAT(m_pTestModule->vNetworks, ElementsAre(m_pTestNetwork));
    EXPECT_THAT(m_pTestModule->vChannels, ElementsAre(m_pTestChan));
}
```
to this:
```cpp
TEST_F(IRCSockTest, OnJoinMessage) {
    CMessage msg(":somebody JOIN #chan");
    m_pTestSock->ReadLine(msg.ToString());

    EXPECT_THAT(m_pTestModule->vsHooks, ElementsAre("OnJoinMessage", "OnSendToClientMessage"));
    EXPECT_THAT(m_pTestModule->vsMessages, ElementsAre(msg.ToString(), msg.ToString()));
    EXPECT_THAT(m_pTestModule->vNetworks, ElementsAre(m_pTestNetwork, m_pTestNetwork));
    EXPECT_THAT(m_pTestModule->vChannels, ElementsAre(m_pTestChan, m_pTestChan));
}
```
Basically, it polluted many of the other tests. Worse, it wasn't super obvious when writing hook tests whether or not they would be affected.

The issue with `HALT` may have been something I was misremembering, as I can't find the relevant failed tests, but the idea is that, when testing some hook that also results in `OnSendToClient/IRCMessage` being called, if the hook action is set to `HALT`, it wasn't necessarily clear if the hook actually being tested would halt execution or if the `OnSend` hook would do so, resulting in incomplete testing.

By adding a field to indicate whether `OnSendToClient/IRCMessage` should process its calls, these issues were avoided.

[onjoinmessage]: https://github.com/elyscape/znc/blob/823ac07/test/IRCSockTest.cpp#L226-L234
